dataplane/netlink.cpp

```diff
@@ -0,0 +1,261 @@
+#include "netlink.hpp"
+
+#include <variant>
+
+#include <net/if.h>
+#include <netlink/netlink.h>
+#include <netlink/route/neighbour.h>
+
+namespace netlink
+{
+
+std::variant<Entry, int> ParseNeighbor(
+        rtnl_neigh* neigh,
+        std::function<std::optional<tInterfaceId>(const char*)> get_id)
+{
+	int sysifid = rtnl_neigh_get_ifindex(neigh);
+	Entry entry;
+	char ifname[IFNAMSIZ];
+	if (if_indextoname(sysifid, ifname) == nullptr)
+	{
+		YANET_LOG_INFO("Skipping message for unknown OS interface '%i'\n", sysifid);
+		return NL_OK;
+	}
+	if (auto id = get_id(ifname); id.has_value())
+	{
+		std::get<tInterfaceId>(entry) = id.value();
+	}
+	else
+	{
+		YANET_LOG_INFO("Skipping message for unconfigured interface '%s'\n", ifname);
+		return NL_OK;
+	}
+
+	nl_addr* oaddr = rtnl_neigh_get_dst(neigh);
+	if (!oaddr)
+	{
+		YANET_LOG_INFO("Skipping message with no destination address\n");
+		return NL_OK;
+	}
+	switch (nl_addr_get_family(oaddr))
+	{
+		case AF_INET:
+		{
+			auto& ip = std::get<ipv6_address_t>(entry);
+			std::fill(std::begin(ip.nap), std::end(ip.nap), 0);
+			ip.mapped_ipv4_address =
+			        ipv4_address_t{*static_cast<uint32_t*>(nl_addr_get_binary_addr(oaddr))};
+			std::get<bool>(entry) = false;
+			break;
+		}
+		case AF_INET6:
+			std::get<ipv6_address_t>(entry).SetBinary(static_cast<uint8_t*>(nl_addr_get_binary_addr(oaddr)));
+			std::get<bool>(entry) = true;
+			break;
+		default:
+			YANET_LOG_INFO("Skipping message with unsupported address family\n");
+			return NL_OK;
+	}
+
+	nl_addr* omac = rtnl_neigh_get_lladdr(neigh);
+	if (omac)
+	{
+		auto mac = static_cast<uint8_t*>(nl_addr_get_binary_addr(omac));
+		std::get<std::optional<rte_ether_addr>>(entry).emplace();
+		std::copy(mac, mac + RTE_ETHER_ADDR_LEN, std::get<std::optional<rte_ether_addr>>(entry).value().addr_bytes);
+	}
+
+	return entry;
+}
+
+std::vector<Entry> Provider::GetHostDump(
+        const std::unordered_map<std::string, tInterfaceId>& ids)
+{
+	rtgenmsg rt_hdr = {.rtgen_family = AF_UNSPEC};
+	nl_sock* sk = nl_socket_alloc();
```

Comment:
(see review comment to netlink.hpp)

we will avoid using goto:
```suggestion
	std::unique_ptr<nl_sock, nl_sock_deleter> sk(nl_socket_alloc());
```